File: services/analytics/src/event-consumer.ts
import { createKafkaConsumer, EventHandler } from '@commercesphere/utils';
import { KAFKA_TOPICS } from '@commercesphere/utils';
import { createLogger } from '@commercesphere/utils';
import { DomainEvent } from '@commercesphere/types';
import { analyticsService } from './analytics.service';
import { config } from './config';

const logger = createLogger({ serviceName: 'analytics-service' });

export class AnalyticsEventConsumer {
  private consumer: ReturnType<typeof createKafkaConsumer> | null = null;

  async start(): Promise<void> {
    try {

      this.consumer = createKafkaConsumer({
        brokers: config.kafka.brokers,
        groupId: 'analytics-service-group',
        clientId: 'analytics-service',
        topics: [KAFKA_TOPICS.ORDERS, KAFKA_TOPICS.PAYMENTS, KAFKA_TOPICS.ANALYTICS],
        fromBeginning: false,
      });


      this.consumer.registerHandler('order.created', this.handleOrderCreated);
      this.consumer.registerHandler('order.completed', this.handleOrderCompleted);
      this.consumer.registerHandler('payment.success', this.handlePaymentSuccess);
      this.consumer.registerHandler('product.viewed', this.handleProductViewed);


      await this.consumer.connect();

      logger.info('Analytics event consumer started', {
        topics: [KAFKA_TOPICS.ORDERS, KAFKA_TOPICS.PAYMENTS, KAFKA_TOPICS.ANALYTICS],
      });
    } catch (error) {
      logger.error('Failed to start analytics event consumer', {
        error: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
  }

  async stop(): Promise<void> {
    if (this.consumer) {
      await this.consumer.disconnect();
      logger.info('Analytics event consumer stopped');
    }
  }

  /**
   * Handle order created event
   */
  private handleOrderCreated: EventHandler = async (event: DomainEvent) => {
    const { orderId, userId, totalAmount, items } = event.payload;

    logger.info('Received order.created event', {
      orderId,
      userId,
      totalAmount,
      eventId: event.id,
    });

    try {

      await analyticsService.updateOrderMetrics(orderId, userId, totalAmount);


      if (items && Array.isArray(items)) {
        for (const item of items) {
          await analyticsService.updateProductMetrics(
            item.productId,
            'purchase',
            item.unitPrice * item.quantity
          );
        }
      }

      logger.info('Order created event processed successfully', { orderId });
    } catch (error) {
      logger.error('Failed to process order created event', {
        orderId,
        error: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
  };

  /**
   * Handle order completed event
   */
  private handleOrderCompleted: EventHandler = async (event: DomainEvent) => {
    const { orderId, userId, totalAmount } = event.payload;

    logger.info('Received order.completed event', {
      orderId,
      userId,
      totalAmount,
      eventId: event.id,
    });

    try {

      await analyticsService.updateOrderMetrics(orderId, userId, totalAmount);

      logger.info('Order completed event processed successfully', { orderId });
    } catch (error) {
      logger.error('Failed to process order completed event', {
        orderId,
        error: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
  };

  /**
   * Handle payment success event
   */
  private handlePaymentSuccess: EventHandler = async (event: DomainEvent) => {
    const { orderId, paymentId, amount } = event.payload;

    logger.info('Received payment.success event', {
      orderId,
      paymentId,
      amount,
      eventId: event.id,
    });

    try {


      logger.info('Payment success event processed successfully', { orderId, paymentId });
    } catch (error) {
      logger.error('Failed to process payment success event', {
        orderId,
        paymentId,
        error: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
  };

  /**
   * Handle product viewed event
   */
  private handleProductViewed: EventHandler = async (event: DomainEvent) => {
    const { productId, userId } = event.payload;

    logger.info('Received product.viewed event', {
      productId,
      userId,
      eventId: event.id,
    });

    try {

      await analyticsService.updateProductMetrics(productId, 'view');

      logger.info('Product viewed event processed successfully', { productId });
    } catch (error) {
      logger.error('Failed to process product viewed event', {
        productId,
        error: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
  };
}

export const analyticsEventConsumer = new AnalyticsEventConsumer();

File: services/recommendation/src/event_consumer.py
import json
import logging
from typing import Dict, Any
from kafka import KafkaConsumer
from uuid import UUID
from .config import settings
from .recommendation_service import RecommendationService

logger = logging.getLogger(__name__)


class EventConsumer:
    """Kafka event consumer for recommendation service"""
    
    def __init__(self):
        self.consumer = None
        self.recommendation_service = RecommendationService()
        self.running = False
    
    def start(self):
        """Start consuming events"""
        try:
            self.consumer = KafkaConsumer(
                'orders',
                'products',
                bootstrap_servers=settings.kafka_broker_list,
                group_id=settings.kafka_group_id,
                value_deserializer=lambda m: json.loads(m.decode('utf-8')),
                auto_offset_reset='earliest',
                enable_auto_commit=True
            )
            
            logger.info("Event consumer started")
            self.running = True
            
            for message in self.consumer:
                if not self.running:
                    break
                
                try:
                    self._process_event(message.topic, message.value)
                except Exception as e:
                    logger.error(f"Failed to process event: {e}")
        
        except Exception as e:
            logger.error(f"Event consumer error: {e}")
            raise
    
    def _process_event(self, topic: str, event: Dict[str, Any]):
        """Process an event based on its type"""
        event_type = event.get('type')
        payload = event.get('payload', {})
        
        logger.info(f"Processing event: {event_type} from topic: {topic}")
        
        if event_type == 'product.viewed':
            self._handle_product_viewed(payload)
        elif event_type == 'order.completed':
            self._handle_order_completed(payload)
        else:
            logger.debug(f"Ignoring event type: {event_type}")
    
    def _handle_product_viewed(self, payload: Dict[str, Any]):
        """Handle product viewed event"""
        try:
            user_id = UUID(payload.get('userId'))
            product_id = UUID(payload.get('productId'))
            
            self.recommendation_service.track_view(user_id, product_id)
            logger.info(f"Processed product.viewed: user={user_id}, product={product_id}")
        except Exception as e:
            logger.error(f"Failed to handle product.viewed event: {e}")
    
    def _handle_order_completed(self, payload: Dict[str, Any]):
        """Handle order completed event"""
        try:
            user_id = UUID(payload.get('userId'))
            items = payload.get('items', [])
            
            for item in items:
                product_id = UUID(item.get('productId'))
                self.recommendation_service.record_purchase(user_id, product_id)
            
            logger.info(f"Processed order.completed: user={user_id}, items={len(items)}")
        except Exception as e:
            logger.error(f"Failed to handle order.completed event: {e}")
    
    def stop(self):
        """Stop consuming events"""
        self.running = False
        if self.consumer:
            self.consumer.close()
            logger.info("Event consumer stopped")
